rosbag2_cpp/src/rosbag2_cpp/cache/cache_consumer.cpp

```diff
@@ -53,11 +55,8 @@ void CacheConsumer::close()
 void CacheConsumer::change_consume_callback(
   CacheConsumer::consume_callback_function_t consume_callback)
 {
+  std::lock_guard<std::mutex> lg(consumer_mutex_);
```

Comment:
one worry I have is that this could affect performance as it effectively stops the thread which is used to process the subscriber queue for possibly quite long. In stress cases (at the edge of I/O capability to write), the cache buffer being flushed can be as much as 500 MB of data if --max-cache-size is set to such value (which is reasonable with fast drives). Then we need to wait for a 500 MB data write. I suspect this could cause message loss due to subscription queue (size 10 for example) not being consumed and having to overwrite messages.

I could test this PR with our benchmarks to see if it makes such a big difference.